### ISSUE TITLE
Co-locate base and view tablets when the partition keys match

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#17043]
This issue is about what we lost by disabling "self-pairing" in materialized views, and what we can do to get it back.

In commit 300e549267b2f51d0afc25b11707ca1d23b441a9 we disabled "**self-pairing**" in materialized views when using tablets. Self-pairing was never properly justified and was even more wrong with tablets, so we disabled it.

Unfortunately, the main downside of disabling self-pairing (and the reason why it was enabled in the first place) is that if one of the nodes is down, without self-pairing consistency or durability guarantees in the **view** are diminished: Consider the following scenario:
1. We have 3 nodes, RF=3.
2. Take node 3 is down.
3. Write to the base table.

With self-pairing, each base replica writes to the view replica on the same node:
* Eventually, two base replicas are successfully written (the third is down).
* Each of these base replicas writes to the view replica on the same node - so eventually two view replicas are successfully written.

Without self-pairing, e.g. node 1 writes a view update to node 2, 2 writes to 3, 3 writes to 1:
* Eventually, two base replicas are successfully written (the third is down).
* One of them tries to send the view update to the dead node, and fails. So only **one** view replica  is successfully written, not two as in the self-pairing case.

It seems like the view's consistency or durability are therefore reduced by disabling self-pairing, but it's not clear what guarantee exactly can be said to be broken, because:
1. Durability is not really reduced - although we have just one copy of the data (instead of two) in the view replica, we do have a second copy as a _view hint_ in the base replica that is paired to the dead view replica. This copy is durable, and will be restored when the dead node comes back.
2. Consistency is not really reduced: The consistency setting read=ONE,write=ALL doesn't work anyway (one node is down so write=ALL won't work), and the consistency setting read=write=QUORUM will continue to work: The read quorum will contain the two live nodes, and since one of them is up to date, the read will see the up-to-date data.

Nevertheless, this situation is not pretty. Users might (wrongly?) expect that even if not guaranteed when you don't write with ALL, "usually" reads with CL=1 "eventually" work correctly no matter which CL you used for the writes. Alternator makes exactly this assumption - eventually-consistent reads are done with CL=1. But now this will no longer be true on the view - since only one copy is written between the two live nodes, half the CL=1 reads will miss the new data. 

Exactly this expectation caused the dtest `materialized_views_test.py`::`TestMaterializedViews`::`test_repair_mv` to fail with tablets (so without self-pairing): That test has the scenario above (3 nodes, one down, RF=3), it writes to the view and then reads with CL=ONE - and is surprised when it doesn't see the new data because the new data is on just one of the two live view replicas, not on both - and waiting and waiting will not solve it.

To return self-pairing, but not inconsistently as it was before 300e549267b2f51d0afc25b11707ca1d23b441a9, we suggest to add a feature where **the base and view replicas will be co-located**. The idea is to have _exactly_ the same tablet map for the base and view table - exactly the same tablets and in the same order. The tablet load balancer will need to know about this restriction, and always move a base tablet and view tablet together (its decision of what to move should be based on the sum of the base table and all its local views). Similarly, whenever a base tablet is split or merged, the same thing should be done to the local views' tablets.

Another benefit of base and view tablet co-location is performance for LSIs (when the view key is the same as the base's): It will ensure that view updates are always local - as we had in the past with vnodes, and therefore perform better.

On the other hand, co-location of base and view tablet co-location can hurt performance when the base and the view have very different sizes or read frequency, so they could have benefited from being split into a different number of tablets. So if we do add tablet co-location, we should perhaps add an option to disable it.

**Comment by nyh** [comment]
The dtest `materialized_views_test.py::TestMaterializedViews::test_complex_repair` shows how the problem described above (which happens without self-pairing) can be taken to an absurd level:

That test has a cluster of 5 nodes, RF=3. Two nodes are stopped. It then writes data with CL=1 which succeeds on the base: Because only two nodes are dead, at least one base replica is alive. Buf if only one base replica is alive, and it is paired view replica is in a dead node, we end up with the data not written at all in the view (just in view hints), and it's not readable in the view while the two nodes are down - even if we wait.

I think that this example just shows that some of the expectations of these tests - and specifically that "test_complex_repair" test, are unrealistic and we don't need to keep these tests working.

**Comment by dorlaor** [comment]
Without opening a can of worms, we may have ignored a major common topology
setting - zones (or racks). In the cloud, almost always (but we do support
single zone), we
deploy the replicas in 3 zones. The failure scenarios are also modeled by
zones:
 - Ok to lose any node
 - Ok to lose a complete zone
 - Not ok to lose > 1 node in > 1 zone since we'll have
availability problem (Quorum).

Now what about placement of views? Seems like we ignored the zones
completely.
Maybe it's time to reconsider the view replica placement and always choose
nodes
within the same zone. Sometimes they'll reside on other nodes, sometimes on
the same
base.

**Comment by dorlaor** [comment]
> @dorlaor this is a good point. You're suggesting that maybe we don't need to enforce that base and view tablet replicas be exactly the same node, as long as they are in the same rack (availability zone). This is probably harder to define and enforce than the simpler "both tables should have the same tablet map" that I suggested above.

I also claim that on a 3 node cluster, with zonal deployment, the view replica should be placed within the same base replica node, for HA purposes.

**Comment by nyh** [comment]
@dorlaor opened an issue https://github.com/scylladb/scylladb/issues/17147 on your rack-aware base-view pairing idea. It's a pre-existing problem that already existed for vnodes, but I agree it's a good opportunity to fix it for tablets. This idea in **this** issue (co-locating base and view tablets on the same node) will solve #17147 automatically - if the base and view replicas are on exactly the same node, certainly they are in the same rack.

**Comment by nyh** [comment]
@avikivity this is the idea you raised on today's call. The issue text is a bit messy, beginning with a long justification, but you can read just the last three paragraphs :-)

**Comment by nyh** [comment]
@tgrabiec before we implement this feature, we should probably introduce some sort of "pluggable" logic in the tablet load balancers, so individual components of the code (e.g., MV) can introduce additional restrictions or preferences to the placement of tablets of different tables.

**Comment by mykaul** [comment]
IIRC, @avikivity told me we do not wish to get self pairing back.

**Comment by mykaul** [comment]
Closing for the time being - we've agreed (I think) we do not wish to have the self pairing. Perhaps as a future optimization...?

**Comment by mykaul** [comment]
> Closing for the time being - we've agreed (I think) we do not wish to have the self pairing. Perhaps as a future optimization...?

Re-opening, I'll leave it on the backlog.

**Comment by nyh** [comment]
> Closing for the time being - we've agreed (I think) we do not wish to have the self pairing. Perhaps as a future optimization...?

Yes, it's an idea for a future optimization. The issue gives an historic overview, and ends with a specific request of adding an option that modifies the tablet load balancer to always move base and view tablets (when those have the same partition key) together. I don't think we should close it even if we don't intend to do it soon or even this year. You can add whatever tags/flags you want to make you not see this issue in your typical searches for scheduled items.

**Comment by swasik** [comment]
> Yes, it's an idea for a future optimization. (...) You can add whatever tags/flags you want to make you not see this issue in your typical searches for scheduled items.

Adding P3 label here as this is the normal priority of items that are on our backlog.

**Comment by piodul** [comment]
> ends with a specific request of adding an option that modifies the tablet load balancer to always move base and view tablets (when those have the same partition key) together.

We will most likely implement this for the sake of CDC with tablets (https://github.com/scylladb/scylladb/issues/16317), it's a part of the design.
